cmd/argo/commands/delete.go

```diff
@@ -2,88 +2,71 @@ package commands
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
 )
 
 // NewDeleteCommand returns a new instance of an `argo delete` command
 func NewDeleteCommand() *cobra.Command {
 	var (
-		selector  string
-		all       bool
-		completed bool
-		older     string
+		flags         listFlags
+		all           bool
+		allNamespaces bool
+		dryRun        bool
 	)
-
 	var command = &cobra.Command{
-		Use: "delete WORKFLOW...",
+		Use:   "delete [--dry-run] [WORKFLOW...|[--all] [--older] [--completed] [--prefix PREFIX] [--selector SELECTOR]]",
+		Short: "delete workflows",
 		Run: func(cmd *cobra.Command, args []string) {
 			ctx, apiClient := client.NewAPIClient()
 			serviceClient := apiClient.NewWorkflowServiceClient()
-			namespace := client.Namespace()
-			var workflowsToDelete []metav1.ObjectMeta
+			var workflows wfv1.Workflows
+			if !allNamespaces {
+				flags.namespace = client.Namespace()
+			}
 			for _, name := range args {
-				workflowsToDelete = append(workflowsToDelete, metav1.ObjectMeta{
-					Name:      name,
-					Namespace: namespace,
+				workflows = append(workflows, wfv1.Workflow{
+					ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: flags.namespace},
 				})
 			}
-			if all || completed || older != "" {
-				// all is effectively the default, completed takes precedence over all
-				if completed {
-					if selector != "" {
-						selector = selector + "," + completedLabelSelector
-					} else {
-						selector = completedLabelSelector
-					}
-				}
-				// you can mix older with either of these
-				var olderTime *time.Time
-				if older != "" {
-					var err error
-					olderTime, err = argotime.ParseSince(older)
-					errors.CheckError(err)
-				}
-				list, err := serviceClient.ListWorkflows(ctx, &workflowpkg.WorkflowListRequest{
-					Namespace:   namespace,
-					ListOptions: &metav1.ListOptions{LabelSelector: selector},
-				})
+			if all || flags.completed || flags.prefix != "" || flags.labels != "" {
+				listed, err := listWorkflows(ctx, serviceClient, flags)
 				errors.CheckError(err)
-				for _, wf := range list.Items {
-					if olderTime != nil && (wf.Status.FinishedAt.IsZero() || wf.Status.FinishedAt.After(*olderTime)) {
+				workflows = append(workflows, listed...)
+			}
+			for _, wf := range workflows {
+				if !dryRun {
+					_, err := serviceClient.DeleteWorkflow(ctx, &workflowpkg.WorkflowDeleteRequest{Name: wf.Name, Namespace: wf.Namespace})
+					if err != nil && status.Code(err) == codes.NotFound {
+						fmt.Printf("Workflow '%s' not found\n", wf.Name)
 						continue
 					}
-					workflowsToDelete = append(workflowsToDelete, wf.ObjectMeta)
-				}
-			}
-			for _, md := range workflowsToDelete {
-				_, err := serviceClient.DeleteWorkflow(ctx, &workflowpkg.WorkflowDeleteRequest{Name: md.Name, Namespace: md.Namespace})
-				if err != nil && apierr.IsNotFound(err) {
-					fmt.Printf("Workflow '%s' not found\n", md.Name)
-					continue
+					log.Debug(reflect.TypeOf(err))
```

Comment:
Do we need to use reflect here?

Reply:
ugh - no - the is pre-existing - but I will fix

Reply:
fixed